### PR TITLE
[whisper] small changes for faster tests

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

whisper has a very slow set of tests. 

This PR applies low-hanging fruit changes to get faster tests (slow tests: 7m47s -> 6m59s on my machine)
- the main dataset is now loaded once per test suite (as opposed to once per test in relevant tests)
- generation tests happen on GPU when possible

### CODE REVIEW

File: tests/models/whisper/test_modeling_whisper.py [hunk 1]
```diff
@@ -1432,32 +1431,16 @@ def test_generate_compilation_all_outputs(self):
 @require_torch
 @require_torchaudio
 class WhisperModelIntegrationTests(unittest.TestCase):
-    def setUp(self):
-        self._unpatched_generation_mixin_generate = transformers.GenerationMixin.generate
```
Comment (gante) [inline tests/models/whisper/test_modeling_whisper.py:1436]: this was only used in one test, and that test should use the base `generate`

idk why it was [added in the first place](https://github.com/huggingface/transformers/pull/29225) 🤔

File: tests/models/whisper/test_modeling_whisper.py [hunk 2]
```diff
@@ -1432,32 +1431,16 @@ def test_generate_compilation_all_outputs(self):
...
-
-    def tearDown(self):
-        transformers.GenerationMixin.generate = self._unpatched_generation_mixin_generate
-
-    @cached_property
-    def default_processor(self):
```
Comment (gante) [inline tests/models/whisper/test_modeling_whisper.py:1442]: unused

File: tests/models/whisper/test_modeling_whisper.py [hunk 3]
```diff
@@ -1432,32 +1431,16 @@ def test_generate_compilation_all_outputs(self):
...
-
-    def _load_datasamples(self, num_samples):
+    @classmethod
+    def setUpClass(cls):
+        # Loads 5 samples from the test dataset
+        num_samples = 5
         ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
-        # automatic decoding with librispeech
         speech_samples = ds.sort("id").select(range(num_samples))[:num_samples]["audio"]
+        cls.speech_samples = [x["array"] for x in speech_samples]
```
Comment (ydshieh) [inline tests/models/whisper/test_modeling_whisper.py:1440]: could we delete this in `tearDownClass` 🙏 ?

File: tests/models/whisper/test_modeling_whisper.py [hunk 4]
```diff
@@ -1586,8 +1569,6 @@ def test_large_logits_librispeech(self):
 
     @slow
     def test_tiny_en_generation(self):
-        torch_device = "cpu"
-        set_seed(0)
```
Comment (ydshieh) [inline tests/models/whisper/test_modeling_whisper.py:1590]: could you run with flakefinder to see if we need seed or not?

File: tests/models/whisper/test_modeling_whisper.py (hunk 1 above)
Comment (ydshieh) [reply tests/models/whisper/test_modeling_whisper.py:1436]: #29225 but i have no idea with it.

Comment (ydshieh) [reply tests/models/whisper/test_modeling_whisper.py:1436]: It is linked to #29312, but the change here is only test. So good for me if it works.

**Review by ydshieh — APPROVED** [review]
LGTM, but it would be good to check a few things.

tearDownClass is a good practice to cleanup the stuff.

File: tests/models/whisper/test_modeling_whisper.py (hunk 1 above)
Comment (gante) [reply tests/models/whisper/test_modeling_whisper.py:1436]: That is implicitly tested: the output checked in the test is different if `num_beams` is not respected :)

(just like in all other beam search integration tests: we check the output, which is sensible enough to detect bad usage of flags)

**Comment by gante** [comment]
@ydshieh regarding dataset loading: I've remembered that we had the same issue in the pipeline tests, so I've pushed a new commit which copies the [pattern from the pipeline tests](https://github.com/huggingface/transformers/blob/101b3fa4ea1cfd66e72a73a0f505108454c24cce/tests/pipelines/test_pipelines_audio_classification.py#L48)

The pattern we added before doesn't have a teardown method -- do we need to add it everywhere the pattern is used?

File: tests/models/whisper/test_modeling_whisper.py (hunk 4 above)
Comment (gante) [reply tests/models/whisper/test_modeling_whisper.py:1590]: `RUN_SLOW=1 py.test tests/models/whisper/test_modeling_whisper.py -k test_tiny_en_generation --flake-finder --flake-runs 100` yields no failures

(`set_seed(0)` comes from the original whisper commit. However, AFAIK, Whisper has no random components -- in fact, many output-checking tests in this file don't set a seed)

**Comment by ydshieh** [comment]
@gante Thank you for checking and further improvement.

Regarding `dataset loading`, it's probably minor (for now), so you can proceed to merge.

(Those loaded objects assigned to class attributes will remain forever until the python process exit. At some point, it might be a problem if we have many test classes)

**Review by vasqu — APPROVED** [review]
LGTM, thanks. Was only concerned about the seeds but seems to be a non-issue.
